### PR TITLE
Feature/orb agent delete

### LINKED ISSUE #135: Agent delete endpoint

### PR DESCRIPTION
**Pull request by lterranova**
Resolves #135 #136 #137

### CODE REVIEW

File: fleet/api/logging.go [hunk 1]
```diff
@@ -178,6 +178,22 @@ func (l loggingMiddleware) ValidateAgentGroup(ctx context.Context, token string,
 	return l.svc.ValidateAgentGroup(ctx, token, s)
 }
 
+func (l loggingMiddleware) RemoveAgent(ctx context.Context, owner, id string) (err error) {
```
Comment (dscabral) [inline fleet/api/logging.go:181]: Here, the second param it's token, not owner

File: fleet/api/metrics.go [hunk 2]
```diff
@@ -62,6 +62,10 @@ func (m metricsMiddleware) ValidateAgentGroup(ctx context.Context, token string,
 	return m.svc.ValidateAgentGroup(ctx, token, s)
 }
 
+func (m metricsMiddleware) RemoveAgent(ctx context.Context, owner string, id string) error {
```
Comment (dscabral) [inline fleet/api/metrics.go:65]: Here, the second param it's token, not owner

File: fleet/postgres/agents_test.go [hunk 3]
```diff
@@ -508,6 +508,58 @@ func TestAgentUpdate(t *testing.T) {
...
+		ID      string
+		ownerID string
+		err     error
+	}{
+		"remove existing agent": {
+			ID:      ag.MFThingID,
+			ownerID: ag.MFOwnerID,
+			err:     nil,
+		},
+		"remove a non-existing agent": {
```
Comment (dscabral) [inline fleet/postgres/agents_test.go:550]: Here you need to pass a invalid MFThingID

File: fleet/redis/producer/events.go [hunk 4]
```diff
@@ -17,6 +17,7 @@ const (
 	AgentCreate      = AgentPrefix + "create"
 	AgentGroupPrefix = "agent_group."
 	AgentGroupRemove = AgentGroupPrefix + "remove"
```
Comment (dscabral) [inline fleet/redis/producer/events.go:19]: Why are you creating this redis stream call?

File: fleet/redis/producer/events.go [hunk 5]
```diff
@@ -61,3 +62,18 @@ func (cce createAgentEvent) encode() map[string]interface{} {
 		"operation": AgentCreate,
 	}
 }
+
```
Comment (dscabral) [inline fleet/redis/producer/events.go:65]: I don't think we need to send a message to redis stream

File: fleet/redis/producer/streams.go [hunk 6]
```diff
@@ -93,6 +93,30 @@ func (es eventStore) ValidateAgentGroup(ctx context.Context, token string, s fle
 	return es.svc.ValidateAgentGroup(ctx, token, s)
 }
 
+func (es eventStore) RemoveAgent(ctx context.Context, token, id string) (err error) {
```
Comment (dscabral) [inline fleet/redis/producer/streams.go:96]: Same here

**Review by dscabral — CHANGES_REQUESTED** [review]

File: fleet/mocks/things.go [hunk 7]
```diff
@@ -183,7 +183,7 @@ func (svc *mainfluxThings) ListChannels(context.Context, string, things.PageMeta
 }
 
 func (svc *mainfluxThings) RemoveChannel(context.Context, string, string) error {
-	panic("not implemented")
+	return nil
```
Comment (dscabral) [inline fleet/mocks/things.go:186]: Here has to be implemented the remove function

File: fleet/mocks/agent.go [hunk 8]
```diff
@@ -88,6 +88,16 @@ func (a agentRepositoryMock) RetrieveAllByAgentGroupID(ctx context.Context, owne
 	return []fleet.Agent{}, nil
 }
 
+func (a agentRepositoryMock) Delete(ctx context.Context, owner, thingID string) error {
```
Comment (dscabral) [inline fleet/mocks/agent.go:91]: Use ownerID to maintain the consistency

File: fleet/agents.go [hunk 9]
```diff
@@ -99,6 +101,8 @@ type AgentRepository interface {
 	UpdateAgentByID(ctx context.Context, ownerID string, agent Agent) error
 	// RetrieveByID retrieves the Agent having the provided ID and owner
 	RetrieveByID(ctx context.Context, ownerID string, thingID string) (Agent, error)
+	// Delete an existing agent by owner and id
+	Delete(ctx context.Context, owner string, thingID string) error
```
Comment (dscabral) [inline fleet/agents.go:105]: Use ownerID to maintain the consistency

**Review by dscabral — CHANGES_REQUESTED** [review]

File: fleet/mocks/things.go [hunk 10]
```diff
@@ -182,8 +182,12 @@ func (svc *mainfluxThings) ListChannels(context.Context, string, things.PageMeta
 	panic("not implemented")
 }
 
-func (svc *mainfluxThings) RemoveChannel(context.Context, string, string) error {
-	panic("not implemented")
+func (svc *mainfluxThings) RemoveChannel(_ context.Context, owner string, id string) error {
+	svc.mu.Lock()
+	defer svc.mu.Unlock()
+
+	delete(svc.channels, key(owner, id))
```
Comment (dscabral) [inline fleet/mocks/things.go:189]: Change the delete(svc.channels, key(owner, id)) to delete(svc.channels, id)

**Review by dscabral — CHANGES_REQUESTED** [review]

**Review by dscabral — DISMISSED** [review]

File: fleet/agent_service.go [hunk 11]
```diff
@@ -149,11 +160,11 @@ func (svc fleetService) RemoveAgent(ctx context.Context, token, thingID string)
 	}
 
 	if errT := svc.mfsdk.DeleteThing(res.MFThingID, token); errT != nil {
-		err = errors.Wrap(err, errT)
+		svc.logger.Error("failed to delete thing", zap.Error(errT))
```
Comment (weyrick) [inline fleet/agent_service.go:163]: Sorry forgot to mention this but please include res.MFThingID as a logger variable, like this:
```
svc.logger.Error("failed to delete thing", zap.Error(errT), zap.String('thing_id', res.MFThingID))
```

File: fleet/agent_service.go [hunk 12]
```diff
@@ -149,11 +160,11 @@ func (svc fleetService) RemoveAgent(ctx context.Context, token, thingID string)
...
 	}
 
 	if errT := svc.mfsdk.DeleteChannel(res.MFChannelID, token); errT != nil {
-		err = errors.Wrap(err, errT)
+		svc.logger.Error("failed to delete channel", zap.Error(errT))
```
Comment (weyrick) [inline fleet/agent_service.go:167]: and same here, including channel id

**Review by weyrick — CHANGES_REQUESTED** [review]

**Review by weyrick — APPROVED** [review]
🎉

**Pull request merged**